### PR TITLE
Add `Discrete(dtype)` parameter

### PR DESCRIPTION
**Pull request by JDRanpariya**
# Description

Add support for dtype in Discrete Space.

Fixes # (issue)
#1118 

## Type of change

Please delete options that are not relevant.

- [ ] Documentation only change (no code changed)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: gymnasium/spaces/discrete.py [hunk 1]
```diff
@@ -84,13 +99,13 @@ def sample(self, mask: MaskNDArray | None = None) -> np.int64:
                 np.logical_or(mask == 0, valid_action_mask)
             ), f"All values of a mask should be 0 or 1, actual values: {mask}"
             if np.any(valid_action_mask):
-                return self.start + self.np_random.choice(
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/spaces/discrete.py:87]: You should be able to pass a type to `choice` or use `asdtype(self.dtype)`

File: gymnasium/spaces/discrete.py [hunk 2]
```diff
@@ -84,13 +99,13 @@ def sample(self, mask: MaskNDArray | None = None) -> np.int64:
...
-                    np.where(valid_action_mask)[0]
+                return self.start + self.dtype.type(
+                    self.np_random.choice(np.where(valid_action_mask)[0])
                 )
             else:
                 return self.start
 
-        return self.start + self.np_random.integers(self.n)
+        return self.start + self.dtype.type(self.np_random.integers(self.n))
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/spaces/discrete.py:108]: See comment above

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Could you move the changes then good to go

File: gymnasium/spaces/discrete.py (hunk 1 above)
Comment (JDRanpariya) [reply gymnasium/spaces/discrete.py:87]: unfortunately that returns an integer and it doesn't have `asdtype` or `astype`.

File: gymnasium/spaces/discrete.py (hunk 2 above)
Comment (JDRanpariya) [reply gymnasium/spaces/discrete.py:108]: Here I was able to use `.astype`

File: gymnasium/spaces/discrete.py (hunk 1 above)
Comment (JDRanpariya) [reply gymnasium/spaces/discrete.py:87]: Let me know if you have any other idea, otherwise would have to go with `self.dtype.type()`

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
I forgot to request testing,

```python
@pytest.mark.parametrize("dtype, sample_dtype", [
    (int, np.int64),
    (np.int64, np.int64),
    (np.int32, np.int32),
    (np.uint8, np.uint8),
])
def test_dtype(dtype, sample_dtype):
    space = Discrete(n=5, dtype=dtype, start=2)

    sample = space.sample()
    sample_mask = space.sample(mask=np.array([0, 1, 0, 0, 0], dtype=np.int8))
    print(f'{sample=}, {sample_mask=}')
    print(f'{type(sample)=}, {type(sample_mask)=}')
    assert isinstance(sample, sample_dtype), type(sample)
    assert isinstance(sample_mask, sample_dtype), type(sample_mask)


@pytest.mark.parametrize("dtype", [
    str,
    np.float32,
    np.complex64,
])
def test_dtype_error(dtype):
    with pytest.raises(ValueError, match="Invalid Discrete dtype"):
        Discrete(4, dtype=dtype)
```
Could you add this to `tests/spaces/test_discrete.py`

File: gymnasium/spaces/discrete.py (hunk 1 above)
Comment (pseudo-rnd-thoughts) [reply gymnasium/spaces/discrete.py:87]: no worries, `self.dtype.type` is fine

**Review by JDRanpariya — COMMENTED** [review]
Have added the tests, thanks!

File: tests/spaces/test_discrete.py [hunk 3]
```diff
@@ -32,3 +33,36 @@ def test_sample_mask():
...
+
+@pytest.mark.parametrize(
+    "dtype",
+    [
+        str,
+        np.float32,
+        np.complex64,
+    ],
+)
+def test_dtype_error(dtype):
```
Comment (Kallinteris-Andreas) [inline tests/spaces/test_discrete.py:66]: Also Test dtype equals none

Comment (Kallinteris-Andreas) [reply tests/spaces/test_discrete.py:67]: Yes, feel free to create a PR for multi-discrete testing next.

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
I just realised that we need to update the `contains` to work with other dtype.
@JDRanpariya Could you update `contains` and add to `test_dtype`, checking that the samples are contained within the space. 
It might be worth checking if failure cases as well including `int` for `np.int64` dtype and empty shape arrays

**Comment by JDRanpariya** [comment]
Hmm, there is no test for `contains', 'to_jsonable` and 'from_jsonable`, would have to know the expected behavior and write tests for all three of them. 

Would look into it, shortly. I would like to spend sometime looking at whole code base for a while.

**Review by KumarAnmol1234 — APPROVED** [review]
